core/modules/indexers/tag-indexer.js

```diff
@@ -35,9 +35,27 @@ TagIndexer.prototype.rebuild = function() {
 };
 
 TagIndexer.prototype.update = function(updateDescriptor) {
-	$tw.utils.each(this.subIndexers,function(subIndexer) {
-		subIndexer.update(updateDescriptor);
-	});
+	var newTid = updateDescriptor.new.tiddler;
+	var oldTid = updateDescriptor.old.tiddler;
+	var noOldTid = oldTid === undefined;
+	var noNewTid = newTid === undefined;
+
+	// Index should be kept when there is no new tiddler
+	if (noNewTid) {
+		return;	// early
+	} //  -> or
+	// a) new tiddler has no tags && no old tiddler -> or
+	// b) new tiddler has no tags && old tiddler has no tags
+	var a=((newTid && newTid.fields && newTid.fields.tags === undefined) && noOldTid),
```

Comment:
You can return early whenever possible. Here just 

```js
// a) new tiddler has no tags && no old tiddler -> or
if ((newTid && newTid.fields && newTid.fields.tags === undefined) && noOldTid) return;
// b) new tiddler has no tags && old tiddler has no tags
if ((newTid && newTid.fields && newTid.fields.tags === undefined) && (oldTid && oldTid.fields && oldTid.fields.tags === undefined) return;

Reply:
These micro optimisations do not make sense. I did add this code to improve readability and make it easier to understand the logic.